jni/ffmpeg-0.10.3/android/armeabi/share/man/man1/ffmpeg.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FFMPEG 1"
.TH FFMPEG 1 "2012-10-02" " " " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ffmpeg \- ffmpeg video converter
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
ffmpeg [global options] [[infile options][\fB\-i\fR \fIinfile\fR]]... {[outfile options] \fIoutfile\fR}...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
ffmpeg is a very fast video and audio converter that can also grab from
a live audio/video source. It can also convert between arbitrary sample
rates and resize video on the fly with a high quality polyphase filter.
.PP
ffmpeg reads from an arbitrary number of input \*(L"files\*(R" (which can be regular
files, pipes, network streams, grabbing devices, etc.), specified by the
\&\f(CW\*(C`\-i\*(C'\fR option, and writes to an arbitrary number of output \*(L"files\*(R", which are
specified by a plain output filename. Anything found on the command line which
cannot be interpreted as an option is considered to be an output filename.
.PP
Each input or output file can in principle contain any number of streams of
different types (video/audio/subtitle/attachment/data). Allowed number and/or
types of streams can be limited by the container format. Selecting, which
streams from which inputs go into output, is done either automatically or with
the \f(CW\*(C`\-map\*(C'\fR option (see the Stream selection chapter).
.PP
To refer to input files in options, you must use their indices (0\-based). E.g.
the first input file is \f(CW0\fR, the second is \f(CW1\fR etc. Similarly, streams
within a file are referred to by their indices. E.g. \f(CW\*(C`2:3\*(C'\fR refers to the
fourth stream in the third input file. See also the Stream specifiers chapter.
.PP
As a general rule, options are applied to the next specified
file. Therefore, order is important, and you can have the same
option on the command line multiple times. Each occurrence is
then applied to the next input or output file.
Exceptions from this rule are the global options (e.g. verbosity level),
which should be specified first.
.PP
Do not mix input and output files \*(-- first specify all input files, then all
output files. Also do not mix options which belong to different files. All
options apply \s-1ONLY\s0 to the next input or output file and are reset between files.
.IP "\(bu" 4
To set the video bitrate of the output file to 64kbit/s:
.Sp
.Vb 1
\&        ffmpeg \-i input.avi \-b:v 64k output.avi
.Ve
.IP "\(bu" 4
To force the frame rate of the output file to 24 fps:
.Sp
.Vb 1
\&        ffmpeg \-i input.avi \-r 24 output.avi
.Ve
.IP "\(bu" 4
To force the frame rate of the input file (valid for raw formats only)
to 1 fps and the frame rate of the output file to 24 fps:
.Sp
.Vb 1
\&        ffmpeg \-r 1 \-i input.m2v \-r 24 output.avi
.Ve
.PP
The format option may be needed for raw input files.
.SH "STREAM SELECTION"
.IX Header "STREAM SELECTION"
By default ffmpeg includes only one stream of each type (video, audio, subtitle)
present in the input files and adds them to each output file.  It picks the
\&\*(L"best\*(R" of each based upon the following criteria; for video it is the stream
with the highest resolution, for audio the stream with the most channels, for
subtitle it's the first subtitle stream. In the case where several streams of
the same type rate equally, the lowest numbered stream is chosen.
.PP
You can disable some of those defaults by using \f(CW\*(C`\-vn/\-an/\-sn\*(C'\fR options. For
full manual control, use the \f(CW\*(C`\-map\*(C'\fR option, which disables the defaults just
described.
.SH "OPTIONS"
.IX Header "OPTIONS"
All the numerical options, if not specified otherwise, accept in input
a string representing a number, which may contain one of the
International System number postfixes, for example 'K', 'M', 'G'.
If 'i' is appended after the postfix, powers of 2 are used instead of
powers of 10. The 'B' postfix multiplies the value for 8, and can be
appended after another postfix or used alone. This allows using for
example '\s-1KB\s0', 'MiB', 'G' and 'B' as postfix.
.PP
Options which do not take arguments are boolean options, and set the
corresponding value to true. They can be set to false by prefixing
with \*(L"no\*(R" the option name, for example using \*(L"\-nofoo\*(R" in the
command line will set to false the boolean option with name \*(L"foo\*(R".
.SS "Stream specifiers"
.IX Subsection "Stream specifiers"
Some options are applied per-stream, e.g. bitrate or codec. Stream specifiers
are used to precisely specify which stream(s) does a given option belong to.
.PP
A stream specifier is a string generally appended to the option name and
separated from it by a colon. E.g. \f(CW\*(C`\-codec:a:1 ac3\*(C'\fR option contains
\&\f(CW\*(C`a:1\*(C'\fR stream specifer, which matches the second audio stream. Therefore it
would select the ac3 codec for the second audio stream.
.PP
A stream specifier can match several stream, the option is then applied to all
of them. E.g. the stream specifier in \f(CW\*(C`\-b:a 128k\*(C'\fR matches all audio
streams.
.PP
An empty stream specifier matches all streams, for example \f(CW\*(C`\-codec copy\*(C'\fR
or \f(CW\*(C`\-codec: copy\*(C'\fR would copy all the streams without reencoding.
.PP
Possible forms of stream specifiers are:
.IP "\fIstream_index\fR" 4
.IX Item "stream_index"
Matches the stream with this index. E.g. \f(CW\*(C`\-threads:1 4\*(C'\fR would set the
thread count for the second stream to 4.
.IP "\fIstream_type\fR\fB[:\fR\fIstream_index\fR\fB]\fR" 4
.IX Item "stream_type[:stream_index]"
\&\fIstream_type\fR is one of: 'v' for video, 'a' for audio, 's' for subtitle,
\&'d' for data and 't' for attachments. If \fIstream_index\fR is given, then
matches stream number \fIstream_index\fR of this type. Otherwise matches all
streams of this type.
.IP "\fBp:\fR\fIprogram_id\fR\fB[:\fR\fIstream_index\fR\fB]\fR" 4
.IX Item "p:program_id[:stream_index]"
If \fIstream_index\fR is given, then matches stream number \fIstream_index\fR in
program with id \fIprogram_id\fR. Otherwise matches all streams in this program.
.SS "Generic options"
.IX Subsection "Generic options"
These options are shared amongst the av* tools.
.IP "\fB\-L\fR" 4
.IX Item "-L"
Show license.
.IP "\fB\-h, \-?, \-help, \-\-help\fR" 4
.IX Item "-h, -?, -help, --help"
Show help.
.IP "\fB\-version\fR" 4
.IX Item "-version"
Show version.
.IP "\fB\-formats\fR" 4
.IX Item "-formats"
Show available formats.
.Sp
The fields preceding the format names have the following meanings:
.RS 4
.IP "\fBD\fR" 4
.IX Item "D"
Decoding available
.IP "\fBE\fR" 4
.IX Item "E"
Encoding available
.RE
.RS 4
.RE
.IP "\fB\-codecs\fR" 4
.IX Item "-codecs"
Show available codecs.
.Sp
The fields preceding the codec names have the following meanings:
.RS 4
.IP "\fBD\fR" 4
.IX Item "D"
Decoding available
.IP "\fBE\fR" 4
.IX Item "E"
Encoding available
.IP "\fBV/A/S\fR" 4
.IX Item "V/A/S"
Video/audio/subtitle codec
.IP "\fBS\fR" 4
.IX Item "S"
Codec supports slices
.IP "\fBD\fR" 4
.IX Item "D"
Codec supports direct rendering
.IP "\fBT\fR" 4
.IX Item "T"
Codec can handle input truncated at random locations instead of only at frame boundaries
.RE
.RS 4
.RE
.IP "\fB\-bsfs\fR" 4
.IX Item "-bsfs"
Show available bitstream filters.
.IP "\fB\-protocols\fR" 4
.IX Item "-protocols"
Show available protocols.
.IP "\fB\-filters\fR" 4
.IX Item "-filters"
Show available libavfilter filters.
.IP "\fB\-pix_fmts\fR" 4
.IX Item "-pix_fmts"
Show available pixel formats.
.IP "\fB\-sample_fmts\fR" 4
.IX Item "-sample_fmts"
Show available sample formats.
.IP "\fB\-loglevel\fR \fIloglevel\fR \fB| \-v\fR \fIloglevel\fR" 4
.IX Item "-loglevel loglevel | -v loglevel"
Set the logging level used by the library.
\&\fIloglevel\fR is a number or a string containing one of the following values:
.RS 4
.IP "\fBquiet\fR" 4
.IX Item "quiet"
.PD 0
.IP "\fBpanic\fR" 4
.IX Item "panic"
.IP "\fBfatal\fR" 4
.IX Item "fatal"
.IP "\fBerror\fR" 4
.IX Item "error"
.IP "\fBwarning\fR" 4
.IX Item "warning"
.IP "\fBinfo\fR" 4
.IX Item "info"
.IP "\fBverbose\fR" 4
.IX Item "verbose"
.IP "\fBdebug\fR" 4
.IX Item "debug"
.RE
.RS 4
.PD
.Sp
By default the program logs to stderr, if coloring is supported by the
terminal, colors are used to mark errors and warnings. Log coloring
can be disabled setting the environment variable
\&\fB\s-1AV_LOG_FORCE_NOCOLOR\s0\fR or \fB\s-1NO_COLOR\s0\fR, or can be forced setting
the environment variable \fB\s-1AV_LOG_FORCE_COLOR\s0\fR.
The use of the environment variable \fB\s-1NO_COLOR\s0\fR is deprecated and
will be dropped in a following FFmpeg version.
.RE
.IP "\fB\-report\fR" 4
.IX Item "-report"
Dump full command line and console output to a file named
\&\f(CW\*(C`\f(CIprogram\f(CW\-\f(CIYYYYMMDD\f(CW\-\f(CIHHMMSS\f(CW.log\*(C'\fR in the current
directory.
This file can be useful for bug reports.
It also implies \f(CW\*(C`\-loglevel verbose\*(C'\fR.
.Sp
Note: setting the environment variable \f(CW\*(C`FFREPORT\*(C'\fR to any value has the
same effect.
.SS "AVOptions"
.IX Subsection "AVOptions"
These options are provided directly by the libavformat, libavdevice and
libavcodec libraries. To see the list of available AVOptions, use the
\&\fB\-help\fR option. They are separated into two categories:
.IP "\fBgeneric\fR" 4
.IX Item "generic"
These options can be set for any container, codec or device. Generic options
are listed under AVFormatContext options for containers/devices and under
AVCodecContext options for codecs.
.IP "\fBprivate\fR" 4
.IX Item "private"
These options are specific to the given container, device or codec. Private
options are listed under their corresponding containers/devices/codecs.
.PP
For example to write an ID3v2.3 header instead of a default ID3v2.4 to
an \s-1MP3\s0 file, use the \fBid3v2_version\fR private option of the \s-1MP3\s0
muxer:
.PP
.Vb 1
\&        ffmpeg \-i input.flac \-id3v2_version 3 out.mp3
.Ve
.PP
All codec AVOptions are obviously per-stream, so the chapter on stream
specifiers applies to them
.PP
Note \fB\-nooption\fR syntax cannot be used for boolean AVOptions,
use \fB\-option 0\fR/\fB\-option 1\fR.
.PP
Note2 old undocumented way of specifying per-stream AVOptions by prepending
v/a/s to the options name is now obsolete and will be removed soon.
.SS "Main options"
.IX Subsection "Main options"
.IP "\fB\-f\fR \fIfmt\fR \fB(\fR\fIinput/output\fR\fB)\fR" 4
.IX Item "-f fmt (input/output)"
Force input or output file format. The format is normally auto detected for input
files and guessed from file extension for output files, so this option is not
needed in most cases.
.IP "\fB\-i\fR \fIfilename\fR \fB(\fR\fIinput\fR\fB)\fR" 4
.IX Item "-i filename (input)"
input file name
.IP "\fB\-y (\fR\fIglobal\fR\fB)\fR" 4
.IX Item "-y (global)"
Overwrite output files without asking.
.IP "\fB\-n (\fR\fIglobal\fR\fB)\fR" 4
.IX Item "-n (global)"
Do not overwrite output files but exit if file exists.
.IP "\fB\-c[:\fR\fIstream_specifier\fR\fB]\fR \fIcodec\fR \fB(\fR\fIinput/output,per\-stream\fR\fB)\fR" 4
.IX Item "-c[:stream_specifier] codec (input/output,per-stream)"
.PD 0
.IP "\fB\-codec[:\fR\fIstream_specifier\fR\fB]\fR \fIcodec\fR \fB(\fR\fIinput/output,per\-stream\fR\fB)\fR" 4
.IX Item "-codec[:stream_specifier] codec (input/output,per-stream)"
.PD
Select an encoder (when used before an output file) or a decoder (when used
before an input file) for one or more streams. \fIcodec\fR is the name of a
decoder/encoder or a special value \f(CW\*(C`copy\*(C'\fR (output only) to indicate that
the stream is not to be re-encoded.
.Sp
For example
.Sp
.Vb 1
\&        ffmpeg \-i INPUT \-map 0 \-c:v libx264 \-c:a copy OUTPUT
.Ve
.Sp
encodes all video streams with libx264 and copies all audio streams.
.Sp
For each stream, the last matching \f(CW\*(C`c\*(C'\fR option is applied, so
.Sp
.Vb 1
\&        ffmpeg \-i INPUT \-map 0 \-c copy \-c:v:1 libx264 \-c:a:137 libvorbis OUTPUT
.Ve
.Sp
will copy all the streams except the second video, which will be encoded with
libx264, and the 138th audio, which will be encoded with libvorbis.
.IP "\fB\-t\fR \fIduration\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-t duration (output)"
Stop writing the output after its duration reaches \fIduration\fR.
\&\fIduration\fR may be a number in seconds, or in \f(CW\*(C`hh:mm:ss[.xxx]\*(C'\fR form.
.IP "\fB\-fs\fR \fIlimit_size\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-fs limit_size (output)"
Set the file size limit.
.IP "\fB\-ss\fR \fIposition\fR \fB(\fR\fIinput/output\fR\fB)\fR" 4
.IX Item "-ss position (input/output)"
When used as an input option (before \f(CW\*(C`\-i\*(C'\fR), seeks in this input file to
\&\fIposition\fR. When used as an output option (before an output filename),
decodes but discards input until the timestamps reach \fIposition\fR. This is
slower, but more accurate.
.Sp
\&\fIposition\fR may be either in seconds or in \f(CW\*(C`hh:mm:ss[.xxx]\*(C'\fR form.
.IP "\fB\-itsoffset\fR \fIoffset\fR \fB(\fR\fIinput\fR\fB)\fR" 4
.IX Item "-itsoffset offset (input)"
Set the input time offset in seconds.
\&\f(CW\*(C`[\-]hh:mm:ss[.xxx]\*(C'\fR syntax is also supported.
The offset is added to the timestamps of the input files.
Specifying a positive offset means that the corresponding
streams are delayed by \fIoffset\fR seconds.
.IP "\fB\-timestamp\fR \fItime\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-timestamp time (output)"
Set the recording timestamp in the container.
The syntax for \fItime\fR is:
.Sp
.Vb 1
\&        now|([(YYYY\-MM\-DD|YYYYMMDD)[T|t| ]]((HH[:MM[:SS[.m...]]])|(HH[MM[SS[.m...]]]))[Z|z])
.Ve
.Sp
If the value is \*(L"now\*(R" it takes the current time.
Time is local time unless 'Z' or 'z' is appended, in which case it is
interpreted as \s-1UTC\s0.
If the year-month-day part is not specified it takes the current
year-month-day.
.IP "\fB\-metadata[:metadata_specifier]\fR \fIkey\fR\fB=\fR\fIvalue\fR \fB(\fR\fIoutput,per\-metadata\fR\fB)\fR" 4
.IX Item "-metadata[:metadata_specifier] key=value (output,per-metadata)"
Set a metadata key/value pair.
.Sp
An optional \fImetadata_specifier\fR may be given to set metadata
on streams or chapters. See \f(CW\*(C`\-map_metadata\*(C'\fR documentation for
details.
.Sp
This option overrides metadata set with \f(CW\*(C`\-map_metadata\*(C'\fR. It is
also possible to delete metadata by using an empty value.
.Sp
For example, for setting the title in the output file:
.Sp
.Vb 1
\&        ffmpeg \-i in.avi \-metadata title="my title" out.flv
.Ve
.Sp
To set the language of the first audio stream:
.Sp
.Vb 1
\&        ffmpeg \-i INPUT \-metadata:s:a:1 language=eng OUTPUT
.Ve
.IP "\fB\-target\fR \fItype\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-target type (output)"
Specify target file type (\f(CW\*(C`vcd\*(C'\fR, \f(CW\*(C`svcd\*(C'\fR, \f(CW\*(C`dvd\*(C'\fR, \f(CW\*(C`dv\*(C'\fR,
\&\f(CW\*(C`dv50\*(C'\fR). \fItype\fR may be prefixed with \f(CW\*(C`pal\-\*(C'\fR, \f(CW\*(C`ntsc\-\*(C'\fR or
\&\f(CW\*(C`film\-\*(C'\fR to use the corresponding standard. All the format options
(bitrate, codecs, buffer sizes) are then set automatically. You can just type:
.Sp
.Vb 1
\&        ffmpeg \-i myfile.avi \-target vcd /tmp/vcd.mpg
.Ve
.Sp
Nevertheless you can specify additional options as long as you know
they do not conflict with the standard, as in:
.Sp
.Vb 1
\&        ffmpeg \-i myfile.avi \-target vcd \-bf 2 /tmp/vcd.mpg
.Ve
.IP "\fB\-dframes\fR \fInumber\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-dframes number (output)"
Set the number of data frames to record. This is an alias for \f(CW\*(C`\-frames:d\*(C'\fR.
.IP "\fB\-frames[:\fR\fIstream_specifier\fR\fB]\fR \fIframecount\fR \fB(\fR\fIoutput,per\-stream\fR\fB)\fR" 4
.IX Item "-frames[:stream_specifier] framecount (output,per-stream)"
Stop writing to the stream after \fIframecount\fR frames.
.IP "\fB\-q[:\fR\fIstream_specifier\fR\fB]\fR \fIq\fR \fB(\fR\fIoutput,per\-stream\fR\fB)\fR" 4
.IX Item "-q[:stream_specifier] q (output,per-stream)"
.PD 0
.IP "\fB\-qscale[:\fR\fIstream_specifier\fR\fB]\fR \fIq\fR \fB(\fR\fIoutput,per\-stream\fR\fB)\fR" 4
.IX Item "-qscale[:stream_specifier] q (output,per-stream)"
.PD
Use fixed quality scale (\s-1VBR\s0). The meaning of \fIq\fR is
codec-dependent.
.IP "\fB\-filter[:\fR\fIstream_specifier\fR\fB]\fR \fIfilter_graph\fR \fB(\fR\fIoutput,per\-stream\fR\fB)\fR" 4
.IX Item "-filter[:stream_specifier] filter_graph (output,per-stream)"
\&\fIfilter_graph\fR is a description of the filter graph to apply to
the stream. Use \f(CW\*(C`\-filters\*(C'\fR to show all the available filters
(including also sources and sinks).
.IP "\fB\-pre[:\fR\fIstream_specifier\fR\fB]\fR \fIpreset_name\fR \fB(\fR\fIoutput,per\-stream\fR\fB)\fR" 4
.IX Item "-pre[:stream_specifier] preset_name (output,per-stream)"
Specify the preset for matching stream(s).
.IP "\fB\-stats (\fR\fIglobal\fR\fB)\fR" 4
.IX Item "-stats (global)"
Print encoding progress/statistics. On by default.
.IP "\fB\-attach\fR \fIfilename\fR \fB(\fR\fIoutput\fR\fB)\fR" 4
.IX Item "-attach filename (output)"
Add an attachment to the output file. This is supported by a few formats
like Matroska for e.g. fonts used in rendering subtitles. Attachments
are implemented as a specific type of stream, so this option will add
a new stream to the file. It is then possible to use per-stream options
on this stream in the usual way. Attachment streams created with this
option will be created after all the other streams (i.e. those created
with \f(CW\*(C`\-map\*(C'\fR or automatic mappings).
.Sp
Note that for Matroska you also have to set the mimetype metadata tag:
.Sp
.Vb 1

src/com/media/ffmpeg/FFMpeg.java
package com.media.ffmpeg;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

import android.content.Context;
import android.util.Log;

import com.media.ffmpeg.android.FFMpegMovieViewAndroid;
import com.media.ffmpeg.config.FFMpegConfig;


public class FFMpeg {
	
	public static final String[] LIBS = new String[] {
		"ffmpeg",
		"mediaplayer",
		"ffmpeg_jni"	// ffmpeg libs compiled to jni lib
	};
	
	public static final String[] EXTENSIONS = new String[] {
		".mp4", 
		".flv", 
		".avi", 
		".wmv"
	};
	
	private static boolean				sLoaded = false;
	private Thread 						mThread;
	private IFFMpegListener 			mListener;
	
	private FFMpegFile					mInputFile;
	private FFMpegFile					mOutputFile;
	private boolean 					mConverting;
	
    public FFMpeg() throws FFMpegException {
    	
    	if(!loadLibs()) {
    		throw new FFMpegException(FFMpegException.LEVEL_FATAL, "Couldn't load native libs");
    	}
    	
    	
    	
    	
    	//native_avcodec_register_all();
		//native_av_register_all();
		mConverting = false;
    }
    
    /**
     * loads all native libraries
     * @return true if all libraries was loaded, otherwise return false
     */
    private static boolean loadLibs() {
    	if(sLoaded) {
    		return true;
    	}
    	boolean err = false;
    	//first load this special name lib
    	
    	for(int i=0;i<LIBS.length;i++) {
    		try {
    			System.loadLibrary(LIBS[i]);
    		} catch(UnsatisfiedLinkError e) {
    			// fatal error, we can't load some our libs
    			Log.d("FFMpeg", "Couldn't load lib: " + LIBS[i] + " - " + e.getMessage());
    			err = true;
    		}
		}
    	if(!err) {
    		sLoaded = true;
    	}
    	return sLoaded;
    }
    
    public FFMpegUtils getUtils() {
    	return new FFMpegUtils();
    }
    
    public boolean isConverting() {
    	return mConverting;
    }
	
	public void setListener(IFFMpegListener listener) {
		mListener = listener;
	}
	
	public FFMpegFile getOutputFile() {
		return mOutputFile;
	}
	
	public FFMpegFile getInputFile() {
		return mInputFile;
	}
	
    public void init(String inputFile, String outputFile) throws RuntimeException, IOException {
		native_av_init();
		
		mInputFile = setInputFile(inputFile);
		mOutputFile = setOutputFile(outputFile);
	}
    
    public void setConfig(FFMpegConfig config) {
    	setFrameSize(config.resolution[0], config.resolution[1]);
		setAudioChannels(config.audioChannels);
		setAudioRate(config.audioRate);
		setFrameRate(config.frameRate);
		setVideoCodec(config.codec);
		setFrameAspectRatio(config.ratio[0], config.ratio[1]);
		setBitrate(config.bitrate);
		
		native_av_parse_options(new String[] {"ffmpeg", mOutputFile.getFile().getAbsolutePath()});
    }
    
    public FFMpegMovieViewAndroid getMovieView(Context context) {
    	return new FFMpegMovieViewAndroid(context);
    }
    
    public void setBitrate(String bitrate) {
    	native_av_setBitrate("b", bitrate);
    }
    
    public void setFrameAspectRatio(int x, int y) {
    	native_av_setFrameAspectRatio(x, y);
    }
    
    public void setVideoCodec(String codec) {
    	native_av_setVideoCodec(codec);
    }
    
    public void setAudioRate(int rate) {
    	native_av_setAudioRate(rate);
    }
    
    public void setAudioChannels(int channels) {
    	native_av_setAudioChannels(channels);
    }
    
    public void setFrameRate(int rate) {
    	native_av_setFrameRate(rate);
    }
    
    public void setFrameSize(int width, int height) {
    	native_av_setFrameSize(width, height);
    }
    
    public FFMpegFile setInputFile(String filePath) throws IOException {
    	File f = new File(filePath);
    	if(!f.exists()) {
    		throw new FileNotFoundException("File: " + filePath + " doesn't exist");
    	}
    	FFMpegAVFormatContext c = native_av_setInputFile(filePath);
    	return new FFMpegFile(f, c);
    }
    
    public FFMpegFile setOutputFile(String filePath) throws FileNotFoundException {
    	File f = new File(filePath);
    	if(f.exists()) {
    		f.delete();
    	}
    	//FFMpegAVFormatContext c = native_av_setOutputFile(filePath);
    	return new FFMpegFile(f, null);
    }
    
    public void newVideoStream(FFMpegAVFormatContext context) {
    	native_av_newVideoStream(context.pointer);
    }
	
	public void convert() throws RuntimeException {
		mConverting = true;
		if(mListener != null) {
			mListener.onConversionStarted();
		}
		
		native_av_convert();
		
		mConverting = false;
		if(mListener != null) {
			mListener.onConversionCompleted();
		}
	}
	
	public void convertAsync() throws RuntimeException {
		mThread = new Thread() {
			@Override
			public void run() {
				try {
					convert();
				} catch (RuntimeException e) {
					if(mListener != null) {
						mListener.onError(e);
					}
				}
			}
		};
		mThread.start();
	}
	
	public void waitOnEnd() throws InterruptedException {
		if(mThread == null) {
			return;
		}
		mThread.join();
	}
	
	public void release() {
		native_av_release(1);
	}
	
	/**
	 * callback called by native code to inform java about conversion
	 * @param total_size
	 * @param time
	 * @param bitrate
	 */
    private void onReport(double total_size, double time, double bitrate) {
		if(mListener != null) {
			FFMpegReport report = new FFMpegReport();
			report.total_size = total_size;
			report.time = time;
			report.bitrate = bitrate;
			mListener.onConversionProcessing(report);
		}
	}
    
    @Override
    protected void finalize() throws Throwable {
    	Log.d("FFMpeg", "finalizing ffmpeg main class");
    	sLoaded = false;
    }
	
    public interface IFFMpegListener {
    	public void onConversionProcessing(FFMpegReport report);
		public void onConversionStarted();
		public void onConversionCompleted();
		public void onError(Exception e);
	}
	
	private native void native_avcodec_register_all();
	
	//{ "native_avdevice_register_all", "()V", (void*) avdevice_register_all },
	
	//{ "native_avfilter_register_all", "()V", (void*) avfilter_register_all },
	
	private native void native_av_register_all();
	
    private native void native_av_init() throws RuntimeException;
    
    private native FFMpegAVFormatContext native_av_setInputFile(String filePath) throws IOException;
    
    private native FFMpegAVFormatContext native_av_setOutputFile(String filePath) throws IOException;
    
    private native int native_av_setBitrate(String opt, String arg);
    
    private native void native_av_newVideoStream(int pointer);
	
    /**
     * ar
     * @param rate
     */
	private native void native_av_setAudioRate(int rate);
	
	private native void native_av_setAudioChannels(int channels);
	
	private native void native_av_setVideoChannel(int channel);
	
	/**
	 * r
	 * @param rate
	 * @throws RuntimeException
	 */
	private native FFMpegAVRational native_av_setFrameRate(int rate) throws RuntimeException;
	
	/**
	 * ration
	 * @param x
	 * @param y
	 */
	private native void native_av_setFrameAspectRatio(int x, int y);
	
	/**
	 * codec
	 * @param codec
	 */
	private native void native_av_setVideoCodec(String codec);
	
	/**
	 * resolution
	 * @param width
	 * @param height
	 */
	private native void native_av_setFrameSize(int width, int height);
	
	private native void native_av_parse_options(String[] args) throws RuntimeException;
	
	private native void native_av_convert() throws RuntimeException;
	
	private native int native_av_release(int code);
	
}
